flink-python/src/main/java/org/apache/flink/client/python/PythonDependencyUtils.java

```diff
@@ -0,0 +1,277 @@
+		 * @param archivePath The path of the archive file.
+		 * @param targetDir The name of the target directory.
+		 */
+		public void addPythonArchive(String archivePath, @Nullable String targetDir) {
+			Preconditions.checkNotNull(archivePath);
+			if (!internalConfig.contains(PYTHON_ARCHIVES)) {
+				internalConfig.set(PYTHON_ARCHIVES, new HashMap<>());
+			}
+			if (targetDir == null) {
+				targetDir = new File(archivePath).getName();
+			}
+			String fileKey = generateUniqueFileKey(PYTHON_ARCHIVE_PREFIX, archivePath + PARAM_DELIMITER + targetDir);
+			registerCachedFileIfNotExist(archivePath, fileKey);
+			internalConfig.get(PYTHON_ARCHIVES).put(fileKey, targetDir);
+		}
+
+		public void configure(ReadableConfig config) {
+			config.getOptional(PythonOptions.PYTHON_FILES).ifPresent(pyFiles -> {
+				for (String filePath : pyFiles.split(FILE_DELIMITER)) {
+					addPythonFile(filePath);
+				}
+			});
+
+			config.getOptional(PythonOptions.PYTHON_REQUIREMENTS).ifPresent(pyRequirements -> {
+				if (pyRequirements.contains(PARAM_DELIMITER)) {
+					String[] requirementFileAndCache = pyRequirements.split(PARAM_DELIMITER, 2);
+					setPythonRequirements(requirementFileAndCache[0], requirementFileAndCache[1]);
+				} else {
+					setPythonRequirements(pyRequirements);
+				}
+			});
+
+			config.getOptional(PythonOptions.PYTHON_ARCHIVES).ifPresent(pyArchives -> {
+				for (String archive : pyArchives.split(FILE_DELIMITER)) {
+					String archivePath;
+					String targetDir;
+					if (archive.contains(PARAM_DELIMITER)) {
+						String[] filePathAndTargetDir = archive.split(PARAM_DELIMITER, 2);
+						archivePath = filePathAndTargetDir[0];
+						targetDir = filePathAndTargetDir[1];
+					} else {
+						archivePath = archive;
+						targetDir = null;
+					}
+					addPythonArchive(archivePath, targetDir);
+				}
+			});
+
+			config.getOptional(PYTHON_EXECUTABLE).ifPresent(e -> internalConfig.set(PYTHON_EXECUTABLE, e));
+
+			config.getOptional(PYTHON_CLIENT_EXECUTABLE).ifPresent(e -> internalConfig.set(PYTHON_CLIENT_EXECUTABLE, e));
+		}
+
+		private String generateUniqueFileKey(String prefix, String hashString) {
+			MessageDigest messageDigest;
+			try {
+				messageDigest = MessageDigest.getInstance(HASH_ALGORITHM);
+			} catch (NoSuchAlgorithmException e) {
+				throw new RuntimeException(e);
+			}
+			messageDigest.update(hashString.getBytes());
+
+			return String.format("%s_%s", prefix, DatatypeConverter.printHexBinary(messageDigest.digest()));
+		}
+
+		private void registerCachedFileIfNotExist(String filePath, String fileKey) {
+			if (cachedFiles.stream().noneMatch(t -> t.f0.equals(fileKey))) {
+				cachedFiles.add(new Tuple2<>(fileKey, new DistributedCache.DistributedCacheEntry(filePath, false)));
+			}
+		}
+
+		private void removeCachedFileIfExists(String fileKey) {
+			cachedFiles.removeAll(cachedFiles.stream().filter(t -> t.f0.equals(fileKey)).collect(Collectors.toSet()));
+		}
+
+		public Configuration getInternalConfiguration() {
```

Comment:
`Internal` is a little confusing as it doesn't explain what it contains. What about renaming it to one meaningful name, e.g. getConfigWithPythonDependencyOptions